internal/clients/options.go

```diff
@@ -1,10 +1,157 @@
+		return nil
+	}
+
+	statusCodes := make([]int, 0)
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		statusCodes = append(statusCodes, opt.StatusCodes...)
+	}
+
+	var maxRetries int32 = 0
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		if opt.MaxRetries > maxRetries {
+			maxRetries = opt.MaxRetries
+		}
+	}
+
+	var retryDelay time.Duration = 0
+	var maxRetryDelay time.Duration = 0
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		if retryDelay == 0 || opt.RetryDelay < retryDelay {
+			retryDelay = opt.RetryDelay
+		}
+		if opt.MaxRetryDelay > maxRetryDelay {
+			maxRetryDelay = opt.MaxRetryDelay
+		}
+	}
+
+	shouldRetry := func(resp *http.Response, err error) bool {
+		for _, opt := range opts {
+			if opt == nil {
+				continue
+			}
+			if opt.ShouldRetry != nil {
+				if opt.ShouldRetry(resp, err) {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	return &policy.RetryOptions{
+		MaxRetries:    maxRetries,
+		RetryDelay:    retryDelay,
+		MaxRetryDelay: maxRetryDelay,
+		StatusCodes:   statusCodes,
+		ShouldRetry:   shouldRetry,
+	}
+}
+
+// NewRetryOptionsForReadAfterCreate creates a RetryOptions for read-after-create operations.
+func NewRetryOptionsForReadAfterCreate(maxRetries int32) *policy.RetryOptions {
+	statusCodes := make([]int, 0)
+	statusCodes = append(statusCodes, DefaultRetryableStatusCodes...)
+	// Add default read after create values for the default retry configuration.
+	statusCodes = append(statusCodes, DefaultRetryableReadAfterCreateStatusCodes...)
+	return &policy.RetryOptions{
+		MaxRetries:  maxRetries,
+		StatusCodes: statusCodes,
+	}
+}
+
+// NewRetryOptions creates a RetryOptions based on the provided retry.RetryValue.
+func NewRetryOptions(rtry retry.RetryValue) *policy.RetryOptions {
+	if rtry.IsNull() || rtry.IsUnknown() {
+		return nil
+	}
+
+	log.Printf("[DEBUG] Using custom retry configuration")
+	return &policy.RetryOptions{
+		RetryDelay:    rtry.GetIntervalSecondsAsDuration(),
+		MaxRetryDelay: rtry.GetMaxIntervalSecondsAsDuration(),
+		// Set a very high max retries to make sure context deadline is respected.
+		MaxRetries:  math.MaxInt16,
+		StatusCodes: DefaultRetryableStatusCodes,
+		ShouldRetry: func(resp *http.Response, err error) bool {
+			if err != nil {
+				// Check if the error message matches any of the retryable error regexps.
+				for _, re := range rtry.GetErrorMessagesRegex() {
+					if re.MatchString(err.Error()) {
+						return true
+					}
+				}
+			} else {
+				responseErr := runtime.NewResponseError(resp)
+				if responseErr != nil {
+					// Check if the error message matches any of the retryable error regexps.
+					for _, re := range rtry.GetErrorMessagesRegex() {
+						if re.MatchString(responseErr.Error()) {
+							log.Printf("[DEBUG] Retrying request due to error: %s matches regex %s", responseErr.Error(), re.String())
+							return true
+						}
+					}
+				}
+			}
+			return false
+		},
```

Comment:
This can be simplified/improved for readability I think, this way there is only one loop to check the regexps and we return early if the errorMsg is empty.

Comment also still stands regarding structured logging with tflog.


```suggestion
    ShouldRetry: func(resp *http.Response, err error) bool {
      // Get the error message to check against regex patterns,
      // If use the err.Error() string first, else get the response error from the HTTP response.
      var errorMsg string
      if err != nil {
          errorMsg = err.Error()
      } else {
        responseErr := runtime.NewResponseError(resp)
        if responseErr != nil {
            errorMsg = responseErr.Error()
        }
      }

      // Check if the error message matches any of the retryable error regexps
      if errorMsg == "" {
        return false
      }

      for _, re := range rtry.GetErrorMessagesRegex() {
        if re.MatchString(errorMsg) {
            log.Printf("[DEBUG] Retrying request due to error: %s matches regex %s", errorMsg, re.String())
            return true
        }
      }

      return false
    },
```